File: easyai-console/src/services/model-config-service.ts
import type { ModelProviderInfo, ModelProviderConfig, ModelConfigGroup, SaveModelProviderConfigRequest, SaveModelConfigGroupRequest } from '@/types/settings';
import { fetchJson, fetchVoid, JSON_HEADERS } from '@/services/api-client';

const API_BASE = '/api/chat';

export const modelConfigService = {
  /**
   * Get all available model providers from server.
   */
  async getAvailableProviders(): Promise<ModelProviderInfo[]> {
    return fetchJson<ModelProviderInfo[]>(`${API_BASE}/model-providers`);
  },

  /**
   * Get models for a specific provider.
   */
  async getModelsForProvider(providerId: string): Promise<ModelProviderInfo['models']> {
    return fetchJson<ModelProviderInfo['models']>(`${API_BASE}/model-providers/${providerId}/models`);
  },

  /**
   * Get user's saved provider configurations.
   */
  async getUserConfigurations(): Promise<ModelProviderConfig[]> {
    return fetchJson<ModelProviderConfig[]>(`${API_BASE}/model-configs`);
  },

  /**
   * Save a new provider configuration.
   */
  async saveConfiguration(request: SaveModelProviderConfigRequest): Promise<ModelProviderConfig> {
    return fetchJson<ModelProviderConfig>(`${API_BASE}/model-configs`, {
      method: 'POST',
      headers: JSON_HEADERS,
      body: JSON.stringify(request),
    });
  },

  /**
   * Delete a provider configuration.
   */
  async deleteConfiguration(id: string): Promise<void> {
    return fetchVoid(`${API_BASE}/model-configs/${id}`, { method: 'DELETE' });
  },

  // ─── Model Config Groups ─────────────────────────────────────────────────────

  /**
   * Get all model config groups with their member models.
   */
  async getGroups(): Promise<ModelConfigGroup[]> {
    return fetchJson<ModelConfigGroup[]>(`${API_BASE}/model-groups`);
  },

  /**
   * Create a new model config group.
   */
  async saveGroup(request: SaveModelConfigGroupRequest): Promise<ModelConfigGroup> {
    return fetchJson<ModelConfigGroup>(`${API_BASE}/model-groups`, {
      method: 'POST',
      headers: JSON_HEADERS,
      body: JSON.stringify(request),
    });
  },

  /**
   * Update a group's connection settings (cascades to member configs).
   */
  async updateGroup(id: string, request: SaveModelConfigGroupRequest): Promise<ModelConfigGroup> {
    return fetchJson<ModelConfigGroup>(`${API_BASE}/model-groups/${id}`, {
      method: 'PUT',
      headers: JSON_HEADERS,
      body: JSON.stringify(request),
    });
  },

  /**
   * Delete a group and all its member model configs.
   */
  async deleteGroup(id: string): Promise<void> {
    return fetchVoid(`${API_BASE}/model-groups/${id}`, { method: 'DELETE' });
  },
};

File: easyai/easyai-api/src/main/kotlin/com/easy/easyai/api/config/ModelConfigService.kt
package com.easy.easyai.api.config

import com.easy.easyai.api.model.ModelConfigGroup
import com.easy.easyai.api.model.ModelInfo
import com.easy.easyai.api.model.ModelProviderConfig
import com.easy.easyai.api.model.ModelProviderInfo
import com.easy.easyai.api.model.SaveModelConfigGroupRequest
import com.easy.easyai.api.model.SaveModelProviderConfigRequest

/**
 * Service interface for managing model provider configurations.
 * Combines pre-defined provider loading and user configuration management.
 * All methods are suspend functions for fully async operation.
 */
interface ModelConfigService {
    /**
     * Get all available model providers from pre-defined configuration.
     */
    suspend fun getAvailableProviders(userId: String = "system"): List<ModelProviderInfo>

    /**
     * Get a specific provider by ID from pre-defined configuration.
     */
    suspend fun getProviderById(id: String, userId: String = "system"): ModelProviderInfo?

    /**
     * Get models for a specific provider from pre-defined configuration.
     */
    suspend fun getModelsForProvider(providerId: String, userId: String = "system"): List<ModelInfo>

    /**
     * Get user's saved provider configurations.
     */
    suspend fun getUserConfigurations(userId: String = "system"): List<ModelProviderConfig>

    /**
     * Get a user's specific provider configuration by ID.
     */
    suspend fun getUserConfiguration(id: String, userId: String = "system"): ModelProviderConfig?

    /**
     * Save a user's provider configuration.
     */
    suspend fun saveUserConfiguration(request: SaveModelProviderConfigRequest, userId: String = "system"): ModelProviderConfig

    /**
     * Delete a user's provider configuration by ID.
     * @return true if the configuration was found and deleted
     */
    suspend fun deleteUserConfiguration(id: String, userId: String = "system"): Boolean

    // ─── Group operations ────────────────────────────────────────────────────────

    /**
     * Get all model config groups with their member models.
     */
    suspend fun getGroups(userId: String = "system"): List<ModelConfigGroup>

    /**
     * Create a new model config group.
     */
    suspend fun saveGroup(request: SaveModelConfigGroupRequest, userId: String = "system"): ModelConfigGroup

    /**
     * Update a group's connection settings, cascading to all member configs.
     */
    suspend fun updateGroup(id: String, request: SaveModelConfigGroupRequest, userId: String = "system"): ModelConfigGroup

    /**
     * Delete a group and all its member model configs.
     */
    suspend fun deleteGroup(id: String, userId: String = "system"): Boolean
}